test/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java

```diff
@@ -498,11 +519,69 @@ public void controlCharacterAreNoMoreValid_CustomRegex() throws Exception {
         }
     }
 
+    @Test
+    public void passwordHashWithPBKDF2() {
+
+        HudsonPrivateSecurityRealm.PBKDF2PasswordEncoder pbkdf2PasswordEncoder = new HudsonPrivateSecurityRealm.PBKDF2PasswordEncoder();
+        mockStatic(HudsonPrivateSecurityRealm.class);
+        when(getPasswordHeader()).thenReturn("$PBKDF2");
+        assertNotNull(pbkdf2PasswordEncoder.encode("password"));
+        assertTrue(PASSWORD_ENCODER.isPasswordHashed("$PBKDF2" + PASSWORD_HASH_ENCODER.encode("password")));
+    }
+
+    @Test
+    public void passwordHashWithInvalidAgorithm()  {
+        HudsonPrivateSecurityRealm.PBKDF2PasswordEncoder pbkdf2PasswordEncoder = new HudsonPrivateSecurityRealm.PBKDF2PasswordEncoder();
+        mockStatic(HudsonPrivateSecurityRealm.class);
+        SecretKeyFactory secretKeyFactory  = mock(SecretKeyFactory.class);
+        assertThrows(NoSuchAlgorithmException.class, () ->   when(SecretKeyFactory.getInstance("")).thenThrow(NoSuchAlgorithmException.class));
+        pbkdf2PasswordEncoder.encode("password");
+    }
+
+    @Test
+    public void passwordHashWithInvalidKeySpec() throws InvalidKeySpecException {
+        HudsonPrivateSecurityRealm.PBKDF2PasswordEncoder pbkdf2PasswordEncoder = new HudsonPrivateSecurityRealm.PBKDF2PasswordEncoder();
+        mockStatic(HudsonPrivateSecurityRealm.class);
+        SecretKeyFactory secretKeyFactory  = mock(SecretKeyFactory.class);
+        when(secretKeyFactory.generateSecret(any())).thenThrow(InvalidKeySpecException.class);
+        pbkdf2PasswordEncoder.encode("password");
+    }
```

Comment:
shouldn't this be expected to throw an exception given the `when` clause and if so isn't this going to be a failing test?

Reply:
this is bad - `pbkdf2PasswordEncoder.encode()` is then returning as if the password it has encoded is valid - smells like a security issue.

Reply:
modified test cases, throws exception when key spec is not valid please verify